src/util/ALE.sol

```diff
@@ -0,0 +1,459 @@
+    }
+
+    /// @notice Update the helper contract
+    /// @param _market The market we want to update the helper contract for
+    /// @param _helper The helper contract
+    function updateMarketHelper(
+        address _market,
+        address _helper
+    ) external onlyOwner {
+        address oldHelper = address(markets[_market].helper);
+        if (oldHelper != address(0)) {
+            markets[_market].buySellToken.approve(oldHelper, 0);
+            markets[_market].collateral.approve(oldHelper, 0);
+        }
+
+        markets[_market].helper = ITransformHelper(_helper);
+        markets[_market].buySellToken.approve(_helper, type(uint256).max);
+        markets[_market].collateral.approve(_helper, type(uint256).max);
+
+        emit NewHelper(_market, _helper);
+    }
+
+    /// @notice Leverage user position by minting DOLA, buying collateral, deposting into the user escrow and borrow DOLA on behalf to repay the minted DOLA
+    /// @dev Requires user to sign message to permit the contract to borrow DOLA on behalf
+    /// @param _value Amount of DOLA to borrow
+    /// @param _market The market contract
+    /// @param _spender The `allowanceTarget` field from the API response.
+    /// @param _swapCallData The `data` field from the API response.
+    /// @param _permit Permit data
+    /// @param _helperData Optional helper data in case the collateral needs to be transformed
+    /// @param _dbrData Optional data in case the user wants to buy DBR and also withdraw some DOLA
+    function leveragePosition(
+        uint256 _value,
+        address _market,
+        address _spender,
+        bytes calldata _swapCallData,
+        Permit calldata _permit,
+        bytes calldata _helperData,
+        DBRHelper calldata _dbrData
+    ) public payable nonReentrant dolaSupplyUnchanged {
+        if (address(markets[_market].buySellToken) == address(0))
+            revert MarketNotSet(_market);
+
+        IMarket market = IMarket(_market);
+
+        // Mint and approve
+        _mintAndApproveDola(_spender, _value);
+
+        // Call the encoded swap function call on the contract at `swapTarget`,
+        // passing along any ETH attached to this function call to cover protocol fees.
+        (bool success, ) = exchangeProxy.call{value: msg.value}(_swapCallData);
+        if (!success) revert SwapFailed();
+
+        // Actual collateral/buyToken bought
+        uint256 collateralAmount = markets[_market].buySellToken.balanceOf(address(this));
+
+        // If there's a helper contract, the buyToken has to be transformed
+        if (address(markets[_market].helper) != address(0)) {
+            collateralAmount = _convertToCollateral(
+                collateralAmount,
+                _market,
+                _helperData
+            );
+        }
+
+        // Deposit and borrow on behalf
+        market.deposit(msg.sender, collateralAmount);
+
+        _borrowDola(_value, _permit, _dbrData, market);
```

Comment:
UX Suggestion: 
Might you want to enforce a CR check here based on a user supplied threshold? 
Given that the amount of collateral purchased can be different than expected, a user may want to specify a ratio such that they are not brought to the brink of liquidation depending on swap execution.